### PR TITLE
[DEV-11399] Add outlays to Spending over time and separate Awards and Transactions data sources

### PR DESCRIPTION
**Pull request by aguest-kc**
**Description:**
Add outlays to the response of the `/spending_over_time` endpoint. Add a new `spending_level` filter field that specifies which data source to use. Use the AwardSearch Elasticsearch index when the `spending_level` is `awards` instead of the TransactionSearch index.

**Technical details:**
The technical details for the knowledge of other developers. Any detailed caveats or specific deployment steps should be outlined here.

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
2. [x] API documentation updated
3. [x] Necessary PR reviewers:
    - [x] Backend
    - [x] Frontend
5. [x] Frontend impact assessment completed
6. [x] Data validation completed
7. [x] Appropriate Operations ticket(s) created
8. [x] Jira Ticket [DEV-11399](https://federal-spending-transparency.atlassian.net/browse/DEV-11399):
    - [x] Link to this Pull-Request

**Area for explaining above N/A when needed:**
```
4. Matview impact assessment completed
No matviews are affected by this change.
```

### CODE REVIEW

File: usaspending_api/search/v2/views/spending_over_time.py [hunk 1]
```diff
@@ -331,16 +458,32 @@ def post(self, request: Request) -> Response:
...
                 ("group", self.group),
                 ("results", results),
                 (
                     "messages",
-                    get_generic_filters_message(self.original_filters.keys(), [elem["name"] for elem in models]),
+                    [
+                        *get_generic_filters_message(self.original_filters.keys(), [elem["name"] for elem in models]),
+                        (
+                            "The 'subawards' field will be deprecated in the future. "
+                            "Set 'spending_level' to 'subawards' instead. See documentation for more information"
```
Comment (sethstoudenmier) [inline usaspending_api/search/v2/views/spending_over_time.py:479]:
```suggestion
                            "Set 'spending_level' to 'subawards' instead. See documentation for more information."
```

File: usaspending_api/search/v2/views/spending_over_time.py [hunk 2]
```diff
@@ -331,16 +458,32 @@ def post(self, request: Request) -> Response:
                 date_range_type=order_by_cols[-1],
                 columns={"aggregated_amount": "aggregated_amount"},
             )
-        else:
-            results = self.query_elasticsearch_for_prime_awards(time_periods)
+            for result in results:
```
Comment (sethstoudenmier) [inline usaspending_api/search/v2/views/spending_over_time.py:461]: Any particular reason to add this to each response as `None` and not the other Award type fields? The API contract claims that all of these fields are required.

It does also claim the obligation fields are required even though they are not present, but I won't fault you on that because it is outside of the scope of this work.

**Review by sethstoudenmier — CHANGES_REQUESTED** [review]
Codeclimate seems to not be happy. Also some questions about the response object / minor cleanup.

File: usaspending_api/api_contracts/contracts/v2/search/spending_over_time.md [hunk 3]
```diff
@@ -63,6 +71,14 @@ This endpoint returns a list of aggregated award amounts grouped by time period
 + `grant_obligations` (required, number)
 + `direct_payment_obligations` (required, number)
 + `other_obligations` (required, number)
++ `total_outlays` (required, number)
```
[the remark this answers is not in the document]
Comment (aguest-kc) [reply usaspending_api/api_contracts/contracts/v2/search/spending_over_time.md:74]: Nice catch, I've updated the API contract to show this field as `nullable`.

File: usaspending_api/search/v2/views/spending_over_time.py (hunk 1 above)
Comment (aguest-kc) [reply usaspending_api/search/v2/views/spending_over_time.py:479]: Fixed the punctuation.

File: usaspending_api/search/v2/views/spending_over_time.py (hunk 2 above)
Comment (aguest-kc) [reply usaspending_api/search/v2/views/spending_over_time.py]: Updated the Subaward result object to match what is defined in the API contract.

File: usaspending_api/search/v2/views/spending_over_time.py [hunk 4]
```diff
@@ -274,12 +358,50 @@ def build_elasticsearch_result(self, agg_response: AggResponse, time_periods: li
...
+
+        results = []
+        date_buckets = agg_response.group_by_time_period.buckets
+        parsed_bucket = None
+
+        if date_buckets is not None:
+            for bucket in date_buckets:
+                parsed_bucket = self.parse_elasticsearch_bucket(bucket.to_dict())
+
+                if parsed_bucket is not None:
```
Comment (sethstoudenmier) [inline usaspending_api/search/v2/views/spending_over_time.py:397]: Is this conditional necessary? I see you set it to `None` above, but won't `parse_elasticsearch_bucket` always return a dictionary?

File: usaspending_api/api_contracts/contracts/v2/search/spending_over_time.md [hunk 5]
```diff
@@ -63,6 +71,14 @@ This endpoint returns a list of aggregated award amounts grouped by time period
 + `grant_obligations` (required, number)
 + `direct_payment_obligations` (required, number)
 + `other_obligations` (required, number)
++ `total_outlays` (required, number, nullable)
+    The aggregate outlay amount for this time period and the given filters.
++ `contract_outlays` (required, number)
```
Comment (sethstoudenmier) [inline usaspending_api/api_contracts/contracts/v2/search/spending_over_time.md:76]: Two things:
* Aren't the type-specific values in the format `Contract_Outlays`?
* Should these be nullable?

**Review by sethstoudenmier — CHANGES_REQUESTED** [review]
Overall LGTM but left some minor cleanup / questions

File: usaspending_api/search/v2/views/spending_over_time.py [hunk 6]
```diff
@@ -208,22 +256,58 @@ def parse_elasticsearch_bucket(self, bucket: dict) -> dict:
...
+            }
+        else:
+            outlay_dictionary = {
+                "Contract_Outlays": None,
+                "Direct_Outlays": None,
+                "Grant_Outlays": None,
+                "Idv_Outlays": None,
+                "Loan_Outlays": None,
+                "Other_Outlays": None,
+            }
```
Comment (zachflanders-frb) [inline usaspending_api/search/v2/views/spending_over_time.py:277]: You could use the values in the `outlay_map` which has the advantage of not having to repeat these values and worry about keeping them in sync. 

```python 
outlay_dictionary = {v: 0 if self.spending_level == "awards" else None for v in outlay_map.values}
```

**Review by zachflanders-frb — APPROVED** [review]

File: usaspending_api/common/helpers/fiscal_year_helpers.py [hunk 7]
```diff
@@ -193,6 +194,38 @@ def create_full_time_periods(min_date, max_date, group, columns):
...
+        result["aggregated_amount"] = result.get("sub-contract", 0) + result.get("sub-grant", 0)
+
+        result["Contract_Obligations"] = result.get("sub-contract", 0)
+        result["Grant_Obligations"] = result.get("sub-grant", 0)
+
+        result["total_outlays"] = None
+        result["Contract_Outlays"] = None
+        result["Grant_Outlays"] = None
+
+        del result["time_period"]["fy"]
```
[the remark this answers is not in the document]
Comment (aguest-kc) [reply usaspending_api/common/helpers/fiscal_year_helpers.py:220]: I was using `pop()` as a way to remove the `sub-contract` and/or `sub-grant` keys that may not exist, without throwing an exception.

This definitely isn't the most explicit way to go about this. I was trying to avoid scanning each dictionary twice (once to see if the key exists and a second time to try and delete the key), but that minor performance saving is probably offset by making the code harder to understand.

File: usaspending_api/search/v2/views/spending_over_time.py (hunk 4 above)
Comment (aguest-kc) [reply usaspending_api/search/v2/views/spending_over_time.py:397]: Good catch. This was left over from the original, Transaction, code where `parsed_bucket` can be `None`. I've removed this check.

File: usaspending_api/common/helpers/fiscal_year_helpers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (aguest-kc) [reply usaspending_api/common/helpers/fiscal_year_helpers.py:220]: Updated the code with an `if` check to only delete these keys if they are present. This is easier to understand than the `.pop()` code.

File: usaspending_api/api_contracts/contracts/v2/search/spending_over_time.md (hunk 5 above)
Comment (aguest-kc) [reply usaspending_api/api_contracts/contracts/v2/search/spending_over_time.md:76]: Yes to both of those and the obligation fields should be in the same format. I've updated the API contract to reflect the casing in the JSON response.

File: usaspending_api/common/helpers/fiscal_year_helpers.py [hunk 8]
```diff
@@ -193,6 +194,38 @@ def create_full_time_periods(min_date, max_date, group, columns):
...
+        subaward_results: Subaward spending_over_time result for a given time period
+        date_range_type: how the results are split
+            - 'fy', 'quarter', or 'month'
+
+    Returns:
+        List of cleaned Subaward results
+    """
+
+    for result in subaward_results:
+        result["time_period"]["fiscal_year"] = result["time_period"]["fy"]
```
[the remark this answers is not in the document]
Comment (aguest-kc) [reply usaspending_api/common/helpers/fiscal_year_helpers.py:210]: It could go there, but I didn't want to any endpoint-specific code to this helper function since it could be used somewhere else in the future with a different response structure.

File: usaspending_api/search/tests/integration/test_spending_over_time.py
[the remark this answers is not in the document]
Comment (aguest-kc) [reply usaspending_api/search/tests/integration/test_spending_over_time.py:1]: Added a test for awards that returns non-null obligation and outlay data.

File: usaspending_api/search/v2/views/spending_over_time.py (hunk 6 above)
Comment (aguest-kc) [reply usaspending_api/search/v2/views/spending_over_time.py]: Good suggestion, I forgot these field names were available in the `outlay_map` dictionary. Updated the code.

**Review by sethstoudenmier — APPROVED** [review]

**Review by zachflanders-frb — APPROVED** [review]
